python.md

```diff
@@ -0,0 +1,122 @@
+---
+layout: default
+title: rpm.org - Python RPM
+---
+
+# Python RPM
+
+A showcase of the Python RPM on real-life examples.
+
+
+## Querying a single package
+
+Query a single installed package by its name. A Python equivalent for
+`rpm -q hello`:
+
+```
+import rpm
+ts = rpm.TransactionSet()
+mi = ts.dbMatch("name", "hello")
+hdr = next(mi, None)
+if hdr:
+    print(hdr[rpm.RPMTAG_NVRA])
+```
+
+
+## Querying a single package
+
+Again, query a single package but this time print more information. A
+Python equivalent for `rpm -qi hello`:
+
+
+```python
+import rpm
+ts = rpm.TransactionSet()
+mi = ts.dbMatch("name", "hello")
+hdr = next(mi, None)
+if hdr:
+    print(
+        f"Name        : {hdr[rpm.RPMTAG_NAME]}\n"
+        f"Version     : {hdr[rpm.RPMTAG_VERSION]}\n"
+        f"Release     : {hdr[rpm.RPMTAG_RELEASE]}\n"
+        f"Architecture: {hdr[rpm.RPMTAG_ARCH]}\n"
+        f"Install Date: {hdr[rpm.RPMTAG_INSTALLTIME]}\n"
+        f"Group       : {hdr[rpm.RPMTAG_GROUP]}\n"
+        f"Size        : {hdr[rpm.RPMTAG_SIZE]}\n"
+        f"License     : {hdr[rpm.RPMTAG_LICENSE]}\n"
+        f"Signature   : TODO\n"
```

Comment:
What is the correct tag here, please?

Reply:
Strictly speaking I don't think there is "a" correct tag, depends on whether the package is signed with RSA or EdDSA.

But as most packages are signed with RSA, you can probably go with `rpm.RPMTAG_RSAHEADER`

Reply:
Thank you. There is also a `rpm.RPMTAG_SIGPGP` tag which returns a similar value. However they both look something like this:

```
b"\x89\x023\x04\x00\x01\x08\x00\x1d\x16!\x04\x11]\xf9\xae\xf8W\x85>\xe8D]\n\x07'p~\xa1[y\xcc\x05\x02e\xb1\x9c\xc4\x00\n\t\x10\x07'p~\xa1[y\xccF\x12\x10\x00\xab\xf7\x00\xaa0)|\x141\xa5G#!\r\xf80\x1c\xecA$\xc8\xe2\x91\xbae\xa9I\xc8\xbf\xdd\x8f[\xbegN\xb3\x14b\xb1h\xe7\x1e\xa0\x9aaDr\xd6\xd5\x98c\x19[\xdc\xf2\x84G\xb6\xafpz\xd1a\xfaD\x91\xf6\xc2\xf9C\xb5\xc4IB(\xdc#\xe1\xca\xdb:@\x08V\x9f\xdfsbB\xc3\xfe\xe8\x9b!\xba\xad\xc3Zm}d\t\xaa9L1L\xdc\x87g\xd3\x98\xa6\xe3\xb8\x91\x96\xe9\x1b\x1f\x8dv\x02r!\xec+g2\xb7%\x15>.\x06$\x94\x92\xd10\xa7\xa7\xb2\xe1\xa1\x04\xbdp\xf01\x1d\x1ca\xcam4^\xa4\x9b\xe5\r\xbd\xda\x99\x05\xedB\x88\xa5\xf6\xd5\xa9W\xa0\x081A\x90\xd0\xdf2\xdex\xe8\xba\x99\xd4\x01\x03\xb2M\xe1\x05\xb3c\xc0\x00\x91\x84Y\xab\x8eu!*\x0e\x13\x99\x1av\x8f\\e\xdf\xe2\xdf\xfe\x97\xdf\xeby\xfd\x07\xaf\x85wB\x98\x1fz\xfe\xcf\xb8\x83\xb6O\x80_\x11\x14\xac\xa6\x00|\xb5\x1b\xa2\xe3\xc1\x0c\x18\x04\xcaX\xea\x0c\xa7\xa1G\xfb#\x14:gW\x0e\xfe\xf4S\r\xa3\xc3V\x12U:gC\x08\x15\x91Q\xaev\x13Sn\r\xfe \xa7\x0e\x07*\x80\xbe-Q\xde\x8d\xfc\x90\xbf\xbep\xdf\\\x03r\xcfw\x9fW\x06\xd4mo\x87\xf5\xcdc!\x87\x872\xdeD\x12{\x00\xad,\x1f\xf8NY\x94Y\x87,T\x99\xd1TFOE\xd1\xec\x8a(z\xd0\xbf\x19\x82\xcf\xa5_\x0cj\xe52\x86\xb5\x92V\xbf'9\xab\x95\xd8\x96\x84\xc3!\xb1\xe6QX\x8a\xe9\xe2\xb2;\t\x11\x0b\x00T\t\xbbJ|\xf8\x92\xa4\xf1d\xf0+2\x1b_x\xb4\xa9\n?`P|p\x89\xbc\xca\xbc\xee\x1f\xc8\xfb\xc6\xf6\x9a\xac`\x97z\x82\xa2?(z\x91\xe2F\t\xb2\xb1!\xd0\x9c9\xd1\xc3Lo*\xa5\x81~:\x99\xa1f\x13\xf89\xd4\xac\xcc\x96\xd8\xc6Z\xe3>\xc5\xddh\xc9^\x95\xd1pA\xc3\xb1\x0e\xf3\xd0o\xed(\x85\xa9'N\xb9\x9e\xe2\x96\x1e\x9d\xb3G\xeep\xdd\xca\xe1\xea\x94j\r\x0fU\x03\x85\xf3x"
```

I can't figure out how to parse something like this from it

```
Signature   : RSA/SHA256, Thu 25 Jan 2024 12:27:00 AM CET, Key ID 0727707ea15b79cc
```

Reply:
It's legacy, ignore it

Reply:
The signature tags are binary OpenPGP signature blobs. To get something human readable out of them, you need to parse them with something. Python gpgme bindings will do the trick no doubt, but to output something similar as rpm's 'rpm -qi ' does, you'd use rpm's queryformat engine:

`print(h.format('%{rsaheader:pgpsig}'))`

The .format() argument is exactly what you'd put into --queryformat, which can do all sorts of stuff: https://rpm-software-management.github.io/rpm/manual/queryformat.html

Reply:
I worked on this recently, and pysequoia can do this now! https://github.com/wiktor-k/pysequoia/issues/27